### ISSUE TITLE
I set same random seed but get different results, why?

### ISSUE THREAD

**Issue by clw5180** [WongKinYiu/ScaledYOLOv4#136]
cudnn.deterministic = True and cudnn.benchmark = False, and there's no operation like F.interpolation, I don't know why, thanks ! ( I use python3.7, pytorch 1.7 with cudatoolkit=10.2, nvcc=10.2)

**Comment by WongKinYiu** [comment]
what random seeds are fixed in your experiments?

1. weight init
2. data loader
3. data augmentation
4. etc...

**Comment by clw5180** [comment]
sorry I don't know, I just call init_seeds(0) at the begining of training process, I thought everything would be ok.

**Comment by WongKinYiu** [comment]
you could check whether the initial weights and train_batch0.jpg are same.

**Comment by WongKinYiu** [comment]
some function/attribute about random seed.
``` python
torch.manual_seed()
torch.cuda.manual_seed()
np.random.seed()
random.seed()
torch.backends.cudnn.deterministic
```

**Comment by WongKinYiu** [comment]
also, the code has bug.

https://github.com/WongKinYiu/ScaledYOLOv4/blob/yolov4-csp/utils/general.py#L25
change this to
``` python
from utils.torch_utils import init_seeds as init_torch_seeds
from utils.torch_utils import is_parallel
```

https://github.com/WongKinYiu/ScaledYOLOv4/blob/yolov4-csp/utils/general.py#L51
change this to
``` python
init_torch_seeds(seed=seed)
```

**Comment by clw5180** [comment]
> also, the code has bug.
> 
> https://github.com/WongKinYiu/ScaledYOLOv4/blob/yolov4-csp/utils/general.py#L25
> change this to
> 
> ```python
> from utils.torch_utils import init_seeds as init_torch_seeds
> from utils.torch_utils import is_parallel
> ```
> 
> https://github.com/WongKinYiu/ScaledYOLOv4/blob/yolov4-csp/utils/general.py#L51
> change this to
> 
> ```python
> init_torch_seeds(seed=seed)
> ```

Thanks a lot ! I have found this bug before :)  I think the reason is my data_loader's num_workers>1,    and I add this to DataLoader
```
def _init_fn(worker_id):
        np.random.seed(0)
```
Now I'm testing again, and I will check the train_batch0.jpg. You code architecture is similar to ultralytics :)

**Comment by clw5180** [comment]
> you could check whether the initial weights and train_batch0.jpg are same.

train_batch0.jpg is the same now, but still can't reproduce the result. I training for 1 epoch. 
```
def init_seeds(seed=0):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    # Speed-reproducibility tradeoff https://pytorch.org/docs/stable/notes/randomness.html
    if seed == 0:  # slower, more reproducible
        cudnn.deterministic = True
        cudnn.benchmark = False
    else:  # faster, less reproducible
        cudnn.deterministic = False
        cudnn.benchmark = True
```

**Comment by WongKinYiu** [comment]
in data loader, there are random.random and np.random.
and have you check if the initial weights are same or not?

**Comment by clw5180** [comment]
> in data loader, there are random.random and np.random.
> and have you check if the initial weights are same or not?

I use --weights 'yolov4-p5.pt', and augment=False in Dataloader, still cannot reproduce.

**Comment by clw5180** [comment]
> in data loader, there are random.random and np.random.
> and have you check if the initial weights are same or not?

and the training loss can reproduce for a few batches, but training for one epoch training loss becomes a little different.

**Comment by clw5180** [comment]
oh my god! I write --weights 'yolov4-p5.pt' as you wrote in github, but it should be --weights yolov4-p5.pt  ; 
if with the quotation marks, 
‘’‘
pretrained = weights.endswith('.pt')
’‘’
becomes false.....

**Comment by clw5180** [comment]
sorry to disturb you, finally I find the nn.Upsample in yolov4-p5.yaml, and many people say that the back-propagation process of  nn.Upsample will cause the result can't be reproduced.
